### ISSUE TITLE
missing method-ambiguity warning

### ISSUE THREAD

**Issue by yuyichao** [JuliaLang/julia#11407]
This is inspired by a newly poped up warning in `PyCall`

``` julia
julia> type A{K, V}
       end

julia> type B{K, V}
       end

julia> f{K, V}(::A{K, V}, ::B{Any, V}) = 1
f (generic function with 1 method)

julia> f{K, V}(::A{K, V}, ::B{K, Any}) = 2
f (generic function with 2 methods)

julia> g{K, V}(::A{K, V}, ::B{K, Any}) = 2
g (generic function with 1 method)

julia> g{K, V}(::A{K, V}, ::B{Any, V}) = 1
g (generic function with 2 methods)

julia> a = A{Any, Any}()
A{Any,Any}()

julia> b = B{Any, Any}()
B{Any,Any}()

julia> f(a, b)
2

julia> g(a, b)
1
```

**Comment by stevengj** [comment]
Presumably it should give a method-ambiguity warning here?

**Comment by yuyichao** [comment]
Yes

**Comment by yuyichao** [comment]
At least one of the reason for this seems to be `jl_args_morespecific` returning `true`

```julia
julia> type A{K, V}
       end

julia> type B{K, V}
       end

julia> ams(t1, t2) = ccall(:jl_args_morespecific, Cint, (Any, Any), t1, t2) != 0
ams (generic function with 1 method)

julia> K = TypeVar(:K, true)
K

julia> K2 = TypeVar(:K2, true)
K2

julia> V = TypeVar(:V, true)
V

julia> V2 = TypeVar(:V2, true)
V2

julia> ams(Tuple{A{K2, V2}, B{K2, Any}}, Tuple{A{K, V}, B{Any, V}})
true
```

This is caused by `jl_type_match(b, a) == (jl_value_t*)jl_false` being true.

@timholy This seems to be something related to what you want to solve with https://github.com/JuliaLang/julia/commit/bcf5361ec68ba02f4a861b4b7673da438ff0c7b1

@JeffBezanson The version you comment out in https://github.com/JuliaLang/julia/commit/7d0c9e1e6de10ec2d93fe5ce05f30c31206c3e33 doesn't seem to have this particular issue. What' the difference between the two?

**Comment by JeffBezanson** [comment]
closed by #18457 - These now give ambiguity errors.
